toffy/reorg.py

```diff
@@ -0,0 +1,145 @@
+            new_path = os.path.join(output_folder, fov)
+            if os.path.exists(new_path):
+                raise ValueError("The following folder {} already exists in {}. If there are "
+                                 "duplicates in your partial run folders, you'll need to determine"
+                                 " which to keep before merging".format(fov, output_folder))
+            shutil.move(os.path.join(cohort_dir, partial, fov), new_path)
+
+        # remove partial folder
+        shutil.rmtree(os.path.join(cohort_dir, partial))
+
+
+def combine_runs(cohort_dir):
+    """Combines FOVs from different runs together, using the run name as a unique identifier
+
+    Args:
+        cohort_dir (str): path to the directory containing individual run folders"""
+
+    # get all runs
+    run_folders = io_utils.list_folders(cohort_dir)
+
+    # create folder to hold all images
+    output_dir = os.path.join(cohort_dir, 'image_data')
+    os.makedirs(output_dir)
+
+    # loop over each run
+    for run in run_folders:
+        run_path = os.path.join(cohort_dir, run)
+
+        fovs = io_utils.list_folders(run_path)
+        for fov in fovs:
+            shutil.move(os.path.join(run_path, fov),
+                        os.path.join(output_dir, run + '_' + fov))
+
+        shutil.rmtree(run_path)
+
+
+def rename_fov_dirs(json_run_path, default_run_dir, output_run_dir=None):
+    """Renames FOV directories with default_name to have custom_name sourced from the run JSON file
+
+    Args:
+        json_run_path (str): path to the JSON run file which contains the custom name values
+        default_run_dir (str): directory containing default named FOVs
+        output_run_dir (str): directory for renamed FOVs. If None, changed in place
+
+    Raises:
+        KeyError: issue reading keys from the JSON file
+        ValueError: there exist default named directories that are not described in the run file
+        UserWarning: not all custom names from the run file have an existing directory
+
+        """
+
+    io_utils.validate_paths(json_run_path)
+    io_utils.validate_paths(default_run_dir)
+
+    # check that new_dir doesn't already exist
+    if output_run_dir is not None:
+        if os.path.exists(output_run_dir):
+            raise ValueError(f"The new directory supplied already exists: {output_run_dir}")
+
+    with open(json_run_path) as file:
+        run_metadata = json.load(file)
+
+    # check for missing or duplicate fov names
+    run_metadata = rename_missing_fovs(run_metadata)
+    run_metadata = rename_duplicate_fovs(run_metadata)
+
+    # retrieve custom names and number of scans for each fov, construct matching default names
+    fov_scan = dict()
+    for fov in run_metadata.get('fovs', ()):
+        custom_name = fov.get('name')
+        run_order = fov.get('runOrder', -1)
+        scans = fov.get('scanCount', -1)
+        if run_order * scans < 0:
+            raise KeyError(f"Could not locate keys in {run_path}")
```

Comment:
This error feels vague, it would be nice to specify which key is missing here.

Also, if both `run_order` and `scans` are not specified in `json_run_path` for this particular `fov`, this control statement will never be entered, meaning `run_order` and `scans` proceed as `-1`. This doesn't look like the intended functionality, let me know if I missed something.

Reply:
This was code Cami took from @ackagel's watcher code. Was this just defensive coding Adam, or is there a reason to think these values would be missing from the JSON?